lib/src/json_schema/json_schema.dart

```diff
@@ -198,40 +232,65 @@ class JsonSchema {
     }
   }
 
-  Future<JsonSchema> _validateAllPathsAsync() {
+  void _baseResolvePaths() {
     if (_root == this) {
       _schemaAssignments.forEach((assignment) => assignment());
 
       // filter out requests that can be resolved locally.
       final List<RetrievalRequest> requestsToRemove = [];
       for (final retrievalRequest in _retrievalRequests) {
         JsonSchema localSchema;
-        try {
-          localSchema = _getSchemaFromPath(retrievalRequest.schemaUri.toString());
-        } catch (e) {
-          // DO NOTHING: if we couldn't resolve the path locally,
-          // it just means we need to make a request after all
+        // check if the ref is actually a standard schema definition, resolve it locally.
+        if (JsonSchemaVersions.allVersions.contains(retrievalRequest.schemaUri.toString())) {
+          final definitionRef = retrievalRequest.schemaUri.toString();
+          localSchema = JsonSchema.createSchema(getJsonSchemaDefinitionByRef(definitionRef));
+          _addSchemaToRefMap(retrievalRequest.schemaUri.toString(), localSchema);
+        } else {
+          // attempt to get the schema from the existing schema cache.
+          try {
+            localSchema = _getSchemaFromPath(retrievalRequest.schemaUri.toString());
+          } catch (e) {
+            // DO NOTHING: if we couldn't resolve the path locally,
+            // it just means we need to make a request after all
+          }
         }
         if (localSchema != null) {
           requestsToRemove.add(retrievalRequest);
         }
       }
       requestsToRemove.forEach((request) => _retrievalRequests.remove(request));
+    }
+  }
+
+  /// Check for refs that need to be fetched, fetch them, and return the final [JsonSchema].
+  void _resolveAllPathsAsync() {
+    _baseResolvePaths();
 
+    if (_root == this) {
       if (_retrievalRequests.isNotEmpty) {
         Future
-            .wait(_retrievalRequests.map((r) => r.retrievalOperation()))
+            .wait(_retrievalRequests.map((r) => r.asyncRetrievalOperation()))
             .then((_) => _thisCompleter.complete(_getSchemaFromPath('#')));
       } else {
         _thisCompleter.complete(_getSchemaFromPath('#'));
       }
       // _logger.info('Marked $_path complete'); TODO: re-add logger
     }
-    return _thisCompleter.future;
   }
 
-  /// Validate that a given [JsonSchema] conforms to the official JSON Schema spec.
-  Future<JsonSchema> _validateSchemaAsync() {
+  /// Check for refs that need to be fetched, fetch them, and return the final [JsonSchema].
+  void _resolveAllPathsSync() {
+    _baseResolvePaths();
+
+    if (_root == this) {
+      if (_retrievalRequests.isNotEmpty) {
```

Comment:
Can this be 1 conditional with `&&`? Or are there plans to add `else` cases down the road?

Reply:
more cases around ref providers in the next ticket